docker-receive/blobstore/blobstore.go

```diff
@@ -0,0 +1,223 @@
+package blobstore
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/context"
+	storage "github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/registry/storage/driver"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/registry/storage/driver/base"
+	registry "github.com/flynn/flynn/Godeps/_workspace/src/github.com/docker/distribution/registry/storage/driver/factory"
+)
+
+const (
+	DriverName = "blobstore"
+
+	blobstorePrefix = "/docker-receive"
+)
+
+func init() {
+	registry.Register(DriverName, factory{})
+}
+
+// factory implements the registry.StorageDriverFactory interface
+type factory struct{}
+
+func (factory) Create(_ map[string]interface{}) (storage.StorageDriver, error) {
+	return NewDriver(), nil
+}
+
+type baseEmbed struct {
+	base.Base
+}
+
+type Driver struct {
+	baseEmbed
+}
+
+func NewDriver() *Driver {
+	return &Driver{
+		baseEmbed: baseEmbed{
+			Base: base.Base{
+				StorageDriver: &driver{},
+			},
+		},
+	}
+}
+
+// driver implements the storage.StorageDriver interface using the blobstore
+// for storage
+type driver struct{}
+
+func (d *driver) Name() string {
+	return DriverName
+}
+
+func (d *driver) GetContent(ctx context.Context, path string) ([]byte, error) {
+	body, err := d.ReadStream(ctx, path, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()
+	return ioutil.ReadAll(body)
+}
+
+func (d *driver) PutContent(ctx context.Context, path string, content []byte) error {
+	_, err := d.WriteStream(ctx, path, 0, bytes.NewReader(content))
+	return err
+}
+
+func (d *driver) ReadStream(ctx context.Context, path string, offset int64) (io.ReadCloser, error) {
+	req, err := http.NewRequest("GET", d.blobstoreURL(path), nil)
+	if err != nil {
+		return nil, err
+	}
+	if offset > 0 {
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", offset))
+	}
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusPartialContent {
+		res.Body.Close()
+		if res.StatusCode == http.StatusNotFound {
+			return nil, storage.PathNotFoundError{Path: path}
+		}
+		return nil, fmt.Errorf("unexpected HTTP status from blobstore: %s", res.Status)
+	}
+	return res.Body, nil
+}
+
+// readCounter wraps an io.Reader and counts how many bytes are read
+type readCounter struct {
+	r io.Reader
+	n int64
+}
+
+func (r *readCounter) Read(p []byte) (int, error) {
+	n, err := r.r.Read(p)
+	atomic.AddInt64(&r.n, int64(n))
+	return n, err
+}
+
+func (d *driver) WriteStream(ctx context.Context, path string, offset int64, reader io.Reader) (int64, error) {
```

Comment:
Is there a reason we're not using the new `FileWriter` interface? https://github.com/docker/distribution/commit/ff03381d4906d9d90dd92b1ad1050a63147464c7